### PR TITLE
king: compatibility with dec 2020 breaching changes

### PR DESCRIPTION
**Pull request by pilfer-pandex**
After trying to make it work manually, I punted on removing ReOrg. The result is as you see.

PR is draft because it should go in at the same time as @eglaysher's version negotiation so that there will exist some serf binary that the king is compatible with at all times.

Please read https://pastebin.com/0dWC0bte to make sure I got the format changes correct.

### CODE REVIEW

**Comment by eglaysher** [comment]
This is pretty hacky, but it should get us to releasing king haskell soonish. This should be fine as a stopgap, but in the long run, I'd like to see a real removal of `ReOrg`.

**Comment by eglaysher** [comment]
Merged this into my ongoing work trying to implement version negotiation. I still can't actually read pills?

```
urbit-king: ParseErr ["Pill","(,,)","1","[]"] "list terminated with non-null atom"
```

**Comment by pilfer-pandex** [comment]
@eglaysher Do you have more info on that error? If it was a parseNoun error, it should have come with a line number.

**Comment by eglaysher** [comment]
I'm not sure that's right; I don't really recall parseNoun errors ever getting line numbers. But the message is coming from `instance FromNoun a => FromNoun [a]`.

File: pkg/hs/urbit-king/lib/Urbit/Vere/Pier/Types.hs [hunk 1]
```diff
@@ -44,11 +44,14 @@ instance Show Nock where
...
-data Pill = Pill
-  { pBootFormulas   :: ![Nock]
-  , pKernelOvums    :: ![Ev]
-  , pUserspaceOvums :: ![Ev]
-  }
+data Pill
+  = PillIvory [Noun]
+  | PillPill
+    { pName         :: Term
+    , pBootFormulae :: ![Nock]
```
Comment (joemfb) [inline pkg/hs/urbit-king/lib/Urbit/Vere/Pier/Types.hs:51]: This should be `Noun`.

File: pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs [hunk 2]
```diff
@@ -218,9 +218,7 @@ instance Show Entropy where
 data ArvoEv
     = ArvoEvWhom ()   Ship
     | ArvoEvWack ()   Entropy
-    | ArvoEvWarn Path Noun
     | ArvoEvCrud Path Noun
-    | ArvoEvVeer Atom Noun
   deriving (Eq, Ord, Show)
```
Comment (joemfb) [inline pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs:222]: This should be `$%(waif wasp wisp)`, where all three are defined in `arvo.hoon`:

```
+$  waif
  ::  %trim: trim state, spam to all
  ::  %what: update from files
  ::  %whey: produce $mass                    :: XX remove, scry
  ::  %verb: toggle laconicity
  ::
  $%  [%trim p=@ud]
      [%what p=(list (pair path (cask)))]
      [%whey ~]
      [%verb p=(unit ?)]
  ==
+$  wasp
  ::  %crud: reroute $ovum with $goof
  ::  %wack: iterate entropy
  ::  %wyrd: check/record runtime kelvin stack
  ::
  $%  [%crud =goof =ovum]
      [%wack p=@uvJ]
      [%wyrd p=vere]
  ==
+$  wisp
  $%  $>(?(%verb %what) waif)
      $>(?(%wack %wyrd) wasp)
      [%whom p=ship]
  ==
```

All of `$waif` and `$>(%wyrd wasp)` (version negotiation event) are missing; `%crud` is wrong.

**Review by joemfb — CHANGES_REQUESTED** [review]
@pilfer-pandex, sorry for the delay.

I've noted an issue with the pill type, and some missing (and one incorrect) Arvo events. The before/after wire restructuring in your pastebin looks correct.

Since they have not changed in this PR, I have not reviewed the vane event/effect definitions for completeness.

File: pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs (hunk 2 above)
Comment (pilfer-pandex) [reply pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs:223]: Elliot added Wyrd, and I've merged his changes in so this should no longer be an issue. How should waif be handled? What processing needs to be done with them, if any?

File: pkg/hs/urbit-king/lib/Urbit/Vere/Pier/Types.hs (hunk 1 above)
Comment (pilfer-pandex) [reply pkg/hs/urbit-king/lib/Urbit/Vere/Pier/Types.hs:60]: Noun or [Noun]?

File: pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs (hunk 2 above)
Comment (pilfer-pandex) [reply pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs:223]: I've merged in Elliot's work so we have Wyrd now. As for waif, what if anything should I do with these events? It seems like I can fail to support whey and verb without issue, but what are the semantics of trim and what?

File: pkg/hs/urbit-king/lib/Urbit/Vere/Pier/Types.hs [hunk 3]
```diff
@@ -44,6 +44,15 @@ instance Show Nock where
...
 --------------------------------------------------------------------------------
 
+-- A Pill is a pair of [pil_p pil_q], where pil_p is cued and pil_q is an
+-- optional set of userspace ovums.
+--
+-- The cued pil_p is a trel of [mot tag dat], where mot is 0 (version number?),
+-- tag is a cord about the type of pill, and dat is the traditional trel of
+-- [pBootForumlas pKernelOvums pUserspaceOvums].
+--
+-- So what's with pil_q? It looks like it is search for the %into.
```
Comment (joemfb) [inline pkg/hs/urbit-king/lib/Urbit/Vere/Pier/Types.hs:54]: This is just how vere organizes the pill + the filesystem as specified via `-A`. It's a crappy design, and shouldn't be duplicated.

File: pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs (hunk 2 above)
Comment (joemfb) [reply pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs:223]: They're events, not effects, so you'd be generating them, not implementing their semantics. You only *need* them if you'll be a) injecting them or b) parsing events produced elsewhere.

File: pkg/hs/urbit-king/lib/Urbit/Vere/Pier/Types.hs (hunk 1 above)
Comment (joemfb) [reply pkg/hs/urbit-king/lib/Urbit/Vere/Pier/Types.hs:60]: Either is fine, of course, but `[Noun]` would be more correct.

File: pkg/hs/urbit-noun-core/lib/Urbit/Noun/Mug.hs [hunk 4]
```diff
@@ -29,4 +23,16 @@ mugAtom = mugBS . atomBytes
...
+mum :: Word32 -> Word32 -> ByteString -> Word32
+mum syd fal key = go syd 0
+  where
+    go syd 8 = fal
+    go syd i =
+      let haz = murmur3 syd key
+          ham = shiftR haz 31 `xor` (haz .&. 0x7fff_ffff)
+      in if ham /= 0
+        then ham
+        else go (syd + 1) (i + 1)
```
Comment (joemfb) [inline pkg/hs/urbit-noun-core/lib/Urbit/Noun/Mug.hs:38]: This file is now missing its trailing newline.

File: pkg/hs/urbit-king/lib/Urbit/Vere/Pier/Types.hs (hunk 3 above)
Comment (joemfb) [reply pkg/hs/urbit-king/lib/Urbit/Vere/Pier/Types.hs:55]: Bumping my previous feedback: this comment is just describing incidental complexity in vere.

**Review by joemfb — APPROVED** [review]
I've noted a couple minor things that could be cleaned up, but these changes LGTM.

**Pull request merged**